### ISSUE TITLE
Assert ddsrt_avl_is_empty before free the object in ddsi_init.c

### ISSUE THREAD

**Issue by vecenyang** [eclipse-cyclonedds/cyclonedds#1634]
In file **cyclonedds/blob/master/src/core/ddsi/src/ddsi_init.c**, main branch.
from **line 1980** to **line 1982**, the code asserts if the **gv->typelib**, **gv->typedeps** and **gv->typedeps_reverse** are empty.
but these three objects are freed after the assert (from line 1985 to line 1987).
The assert fail only happens when you debug the example.
should we move the assert after the freed code?

**Comment by eboasson** [comment]
I don't think so: calling `ddsrt_avl_free` actually only sets the root to a null pointer if the last argument is a null pointer:
https://github.com/eclipse-cyclonedds/cyclonedds/blob/87b31771ad4dda92afccc6ad1cb84cb7f752b66b/src/ddsrt/src/avl.c#L174-L181

The assert is there to check that this won't leak memory. (Why call `ddsrt_avl_free` in this case? Good question.)

We could of course pass a non-null pointer into `ddsrt_avl_free` and free everything that's still in the type library. Before we do that, I think we should understand *why* it isn't empty in this case. It could be a bug, after all.

It could also be that the example doesn't play by the rules: the types in the library are reference-counted, and if the application manipulates these types directly, then perhaps it forgot to release a reference. For most of the API (like entities such as topics, readers, writers) you never have to worry about that, but we for the type library made a different trade-off.

Can you tell me how to reproduce it?

**Comment by vecenyang** [comment]
sure, just open a terminal (I'm using debian 11) and following the steps:

1. mkdir build
2. cd build
3. cmake -DCMAKE_BUILD_TYPE=Debug -DBUILD_EXAMPLES=True ../
4. make
5. ./bin/dyntype
6. after it require **press enter to exit**, just press Enter.
Then you'll see the **Assertion failed** in the terminal.
